### DISCUSSION TITLE
Altitude techniques

### DISCUSSION THREAD

**Discussion by RhythmOfTension** [pelicanmapping/osgearth#1819] [Support Q&A] [answered]
Hi,

I have a problem understanding the specific uses of different altitude techniques. I understand that drape is best for polygons without extrusion, and maybe lines without tessellation (it helps vs z-fighting). But what about map, scene, gpu? What's the best application for each technique?

Thanks.

**Comment by gwaldron** [comment, marked as answer]
Here is a summary of the various clamping techniques and the various pros and cons.

* **Map** samples the elevation at each vertex and alter the "Z" values to line up with the terrain. This happens once when the feature first loads and compiles. By default this will clamp to the highest elevation data available.

* **Scene** is similar to *Map* but re-clamps the geometry when new terrain data pages in. This is more CPU-intensive and it suitable for smaller localized geometries. It only works on Annotations (not on FeatureModel layers) at the moment.

* **Drape** renders the geometry to a texture, and then uses projective texturing to render that texture on the terrain surface. You don't have to worry about z-flighting, but the resolution of the projected texture is fixed and therefore the technique can result in aliasing artifacts especially at low camera pitch.

* **GPU**: a [somewhat experimental] technique that uses a GPU shader to re-clamp the geometry at rendering time. It uses a depth texture to determine the terrain elevation, and can therefore suffer from precision-related artifacts.

Ways to mitigate z-fighting:

* **Tessellation** of lines helps since the system has more points to clamp to the terrain, reducing the chance of a line segment passing though a terrain triangle. Example: `stroke-tessellation-size: 10km;`

*  **Depth offset** is a shader technique that tweaks the depth buffer value of a geometry so it doesn't render at the exact same depth as the terrain triangle. Example: `render-depth-offset: true;`

*  **Altitude offset** moves each vertex in the Z direction after clamping, creating a separation between geometry and terrain. Example: `altitude-offset: 100;`

Hope this helps!

**Reply by RhythmOfTension** [reply]
Thank you!
